onnxruntime/core/providers/rocm/nn/pool.cc

```diff
@@ -0,0 +1,293 @@
+
+ private:
+  miopenPoolingDescriptor_t desc_;
+};
+
+template <typename T, typename PoolType>
+Status Pool<T, PoolType>::ComputeInternal(OpKernelContext* context) const {
+  typedef typename ToHipType<T>::MappedType HipT;
+  const Tensor* X = context->Input<Tensor>(0);
+  const TensorShape& x_shape = X->Shape();
+  const auto x_dims = x_shape.GetDims();
+
+  if (x_shape.NumDimensions() < 3) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Input dimension cannot be less than 3.");
+  }
+
+  auto kernel_shape = pool_attrs_.kernel_shape;
+  auto pads = pool_attrs_.pads;
+  auto strides = pool_attrs_.strides;
+
+  if (pool_attrs_.global_pooling) {
+    kernel_shape.assign(x_dims.begin() + 2, x_dims.end());
+    pads.assign(kernel_shape.size(), 0);
+    strides.assign(kernel_shape.size(), 1);
+  }
+
+  auto y_dims = pool_attrs_.SetOutputSize(x_shape, x_shape[1], &pads);
+  TensorShape y_shape(y_dims);
+  Tensor* Y = context->Output(0, y_shape);
+  // special case when there is a dim value of 0 in the shape.
+  if (y_shape.Size() == 0)
+    return Status::OK();
+
+  auto x_data = reinterpret_cast<const HipT*>(X->template Data<T>());
+  auto y_data = reinterpret_cast<HipT*>(Y->template MutableData<T>());
+
+  TensorShapeVector x_dims_miopen(x_dims.cbegin(), x_dims.cend());
+  TensorShapeVector y_dims_miopen(y_dims);
+  if (kernel_shape.size() < 2) {
+    // miopen only takes 4D or 5D input, so pad dimensions if needed
+    x_dims_miopen.push_back(1);
+    y_dims_miopen.push_back(1);
+    pads.insert(pads.begin() + kernel_shape.size(), 0);
+    pads.insert(pads.end(), 0);
+    kernel_shape.push_back(1);
+    strides.push_back(1);
+  }
+
+  miopenPoolingMode_t mode = miopenPoolingMax;
+  if constexpr (PoolType::type == onnxruntime::PoolType::kAveragePool) {
+    mode = pool_attrs_.count_include_pad ? miopenPoolingAverageInclusive
+                                         : miopenPoolingAverage;
+  }
+  MiopenPoolingDescriptor pooling_desc;
+  ORT_RETURN_IF_ERROR(pooling_desc.Set(mode, kernel_shape, pads, strides));
+
+  constexpr bool pooling_no_backward = false;
+  if constexpr (std::is_same<T, int8_t>::value || std::is_same<T, uint8_t>::value) {
+    // Cast to float back and forth using temp buffer
+    const auto alpha = Consts<float>::One;
+    const auto beta = Consts<float>::Zero;
+    MiopenTensor x_tensor;
+    MiopenTensor y_tensor;
+    ORT_RETURN_IF_ERROR(x_tensor.Set(x_dims_miopen, MiopenTensor::GetDataType<float>()));
+    ORT_RETURN_IF_ERROR(y_tensor.Set(y_dims_miopen, MiopenTensor::GetDataType<float>()));
+
+    const auto input_count = x_shape.Size();
+    const auto output_count = y_shape.Size();
+
+    IAllocatorUniquePtr<float> temp_X = GetScratchBuffer<float>(input_count);
+    auto temp_Y = GetScratchBuffer<float>(output_count);
+    Impl_Cast<HipT, float>(Stream(), reinterpret_cast<const HipT*>(x_data), temp_X.get(), input_count);
+    MIOPEN_RETURN_IF_ERROR(miopenPoolingForward(MiopenHandle(), pooling_desc,
```

Comment:
I compared this file between CUDA and ROCm version, this is the only difference that cannot be handled by hipify. The ROCm version has 3 more augments (false, nullptr, 0) at the end. I suggest to add a util function named PoolingForwoard (with same augment list as CUDA's version) to both CUDA and ROCm shared util code files, and call it here so we don't need to duplicate this code file for ROCm.

Reply:
Got it, will do it tomorrow.

Reply:
Sorry for the delay. The updated version is ready for review.